modules/reactor/src/meshgenerators/PinMeshGenerator.C

```diff
@@ -690,3 +697,158 @@ PinMeshGenerator::generate()
 
   return std::move(*_build_mesh);
 }
+
+std::unique_ptr<CSG::CSGBase>
+PinMeshGenerator::generateCSG()
+{
+  // Must be called to free the ReactorMeshParams mesh
+  freeReactorParamsCSG();
+
+  auto csg_obj = std::make_unique<CSG::CSGBase>();
+
+  unsigned int radial_index = 0;
+  std::vector<std::vector<std::reference_wrapper<const CSG::CSGSurface>>> surfaces_by_radial_region;
+
+  // Add surfaces corresponding to pin rings
+  for (const auto & radius : _ring_radii)
+  {
+    const auto surf_name = name() + "_radial_ring_" + std::to_string(radial_index);
+    std::unique_ptr<CSG::CSGSurface> ring_surf_ptr =
+        std::make_unique<CSG::CSGZCylinder>(surf_name, 0, 0, radius);
+    const auto & ring_surf = csg_obj->addSurface(std::move(ring_surf_ptr));
+    surfaces_by_radial_region.push_back({ring_surf});
+    ++radial_index;
+  }
+
+  // Add surfaces corresponding to pin ducts
+  for (const auto & duct_halfpitch : _duct_halfpitch)
+  {
+    const auto & duct_surfaces = getOuterRadialSurfaces(radial_index, duct_halfpitch, *csg_obj);
+    surfaces_by_radial_region.push_back(duct_surfaces);
+    ++radial_index;
+  }
+
+  // Add surfaces corresponding to outer pin boundary
+  const auto & duct_surfaces = getOuterRadialSurfaces(radial_index, _pitch / 2., *csg_obj);
+  surfaces_by_radial_region.push_back(duct_surfaces);
+
+  // Define all radial regions
+  std::vector<CSG::CSGRegion> radial_regions;
+  CSG::CSGRegion inner_region, outer_region;
+  for (const auto & radial_surfaces : surfaces_by_radial_region)
+  {
+    CSG::CSGRegion radial_region;
+    if (inner_region.getRegionType() == CSG::CSGRegion::RegionType::EMPTY)
+    {
+      // We are in the innermost radial region, the radial region is inner_region
+      inner_region = CSGUtils::getInnerRegion(radial_surfaces, Point(0, 0, 0));
+      radial_region = inner_region;
+    }
+    else
+    {
+      // For all other regions, the radial region is the intersection of inner_region and
+      // outer_region
+      outer_region = ~inner_region;
+      inner_region = CSGUtils::getInnerRegion(radial_surfaces, Point(0, 0, 0));
+      radial_region = inner_region & outer_region;
+    }
+    radial_regions.push_back(radial_region);
+  }
+
+  // Define all axial surfaces and regions
+  std::vector<CSG::CSGRegion> axial_regions;
+  std::vector<std::reference_wrapper<const CSG::CSGSurface>> surfaces_by_axial_region;
+  if (_extrude)
+  {
+    const auto axial_boundaries = getReactorParam<std::vector<Real>>(RGMB::axial_mesh_sizes);
+    Real axial_level = 0.;
+    for (const auto i : make_range(axial_boundaries.size() + 1))
+    {
+      axial_level += (i != 0) ? axial_boundaries[i - 1] : 0.;
```

Comment:
it needs a real z coordinate we cant just use 0

if the pin extrusion levels are below 0 we d have an empty csg cell

Reply:
I think for the general AdvancedExtruderGenerator that's probably true. For RGMB, the lowest axial level is always assumed to be 0.

Reply:
true true